### ISSUE TITLE
Don't invent new "perm" string format in the yaml

### ISSUE THREAD

**Issue by morrone** [ovis-hpc/ldms#1555]
@nick-enoent I missed this one before it landed. I see the recent change here:

https://github.com/ovis-hpc/ldms/blob/0307e5782940818e9555a01e94ddb7d0df2e3ea2/ldms/python/ldmsd/parser_util.py#L631

I do not think that we should invent a one-off way of representing unix-like permissions in the yaml code. Since ldms is emulating unix file permissions, instead of "rwx-rwx-rwx", one would expect that to look more like "rwxrwxrwx". The minus sign separators seem unnecessarily inconsistent with unix.

And wouldn't the minus signs normally represent a bit that is not set? So wouldn't we expect a user to do "rw-------" for 0600?

OH. Now that I read perm_handler() I see what is going on. The man page does actually explain this now that I take a second read, but the man page is very misleading when it says "Permissions are defined in a string linux file system format". Nothing in linux, that I am aware of, sets permissions using a string like this.

I think we should really cut out this non-standard way of setting permissions.

It might be better to copy the syntax that chmod uses, and everyone is familiar with, for doing the non-numeric representation of setting permission bits. chmod (from the man page) has this syntax:

```
       Each MODE is of the form '[ugoa]*([-+=]([rwxXst]*|[ugo]))+|[-+=][0-7]+'.
```

The yaml could (should?) probably be a subset of that, since ldms doesn't use the Xst bits, etc.

**Comment by tom95858** [comment]
@morrone, @nick-enoent What we were trying accomplish was to make it a string instead of an integer, the format of which was controlled by the YAML parser. Let's propose a format or formats and then implement this once.

**Comment by tom95858** [comment]
@morrone, @nick-enoent let's do it quickly? ;-)

**Comment by tom95858** [comment]
I propose the following:

Unix-like format:
rw-rw-rw-

The 'x' doesn't do anything. If people use it, we could issue a warning that the 'x' is ignored.

Octal Format:

0[0-9]{3}, for example 0644. The leading zero tell us it's octal.

Decimal Format:

Any number absent a leading 0.

[1-9][0-9]*, for example 420, i.e. 0644

But in all cases, the argument is presumed to be a string so that our logic and not the YAML parser does the conversion.

**Comment by tom95858** [comment]
@nick-enoent could you code that up?

**Comment by morrone** [comment]
> I propose the following:
> 
> Unix-like format: rw-rw-rw-
> 
> The 'x' doesn't do anything. If people use it, we could issue a warning that the 'x' is ignored.
> 
> Octal Format:
> 
> 0[0-9]{3}, for example 0644. The leading zero tell us it's octal.
> 
> Decimal Format:
> 
> Any number absent a leading 0.
> 
> [1-9][0-9]*, for example 420, i.e. 0644
> 
> But in all cases, the argument is presumed to be a string so that our logic and not the YAML parser does the conversion.

I don't think you can "presume" that. The yaml spec considers the following numbers and most parsers will just do the conversion to the right native-language numerical type automatically. In other words, these _are numbers_ in yaml:

```
  perm: 0644
  perm: 420
```

I python, we don't even necessarily know that the number was specified in octal or decimal, we just have an integer.

To treat them only as strings, we would just have to raise an error when either of the above was specified, and always require user to force the value to be a string, for instance with quotes:

```
  perm: "0644"
```

 I would suggest:

Allow the field to be either an Integer or a String. If it is an Integer, then assume it is a bit field, sanity check the bits, and pass the number to the lower level as a octal, I think it pretty much this does already, although I don't remember if there is much validation.

If it is a string, the string can either be entirely digits representing an octal value or it can be a unix-like string.

I don't think there is any reason to support non-octal numbers in the string format. I don't think humans use non-octal when working with permission bits.

The suggestion of unix-like format: "rw-rw-rw-", where the characters have a fixed position in the string, is fine with me. To be specific, the allowed pattern would be:

```(r|-)(w|-)-(r|-)(w|-)-(r|-)(w|-)-```

**Comment by tom95858** [comment]
> [...snip...]
> 
> I don't think you can "presume" that. The yaml spec considers the following numbers and most parsers will just do the conversion to the right native-language numerical type automatically. In other words, these _are numbers_ in yaml:
> 

Yes we can, but I think you are saying "should we?"

> ```
>   perm: 0644
>   perm: 420
> ```

Yes, but what happens is that when you publish this as a configuration file with the export option, what does it look like?

> 
> I python, we don't even necessarily know that the number was specified in octal or decimal, we just have an integer.
> 

No true. Python octal is specified with a leading '0o'. That said this is not the most user-friendly.

> To treat them only as strings, we would just have to raise an error when either of the above was specified, and always require user to force the value to be a string, for instance with quotes:
> 
> ```
>   perm: "0644"
> ```
> 
> I would suggest:
> 
> Allow the field to be either an Integer or a String. If it is an Integer, then assume it is a bit field, sanity check the bits, and pass the number to the lower level as a octal, I think it pretty much this does already, although I don't remember if there is much validation.
> 

That's fine, but we need to decide what the configuration file export looks like.

> If it is a string, the string can either be entirely digits representing an octal value or it can be a unix-like string.
> 
> I don't think there is any reason to support non-octal numbers in the string format. I don't think humans use non-octal when working with permission bits.
> 
> The suggestion of unix-like format: "rw-rw-rw-", where the characters have a fixed position in the string, is fine with me. To be specific, the allowed pattern would be:
> 
> `(r|-)(w|-)-(r|-)(w|-)-(r|-)(w|-)-`

Correct...

**Comment by morrone** [comment]
> > I python, we don't even necessarily know that the number was specified in octal or decimal, we just have an integer.
> 
> No true. Python octal is specified with a leading '0o'. That said this is not the most user-friendly.

I think may be confusing data types with automatic string conversions. Octal is not a data type in python. int() is a data type. The generic yaml parsing library will see "yaml number" and convert it into an "int()" in python. LDMS's python code would not 
see "octal" when the the yaml scalar value is a number.

> That's fine, but we need to decide what the configuration file export looks like.

I'm not suggesting any change to that. I think Nick's perm_handler() code always generates the LDMS's octal string representation, and I think it is a good way to handle that.

**Comment by morrone** [comment]
Oh, actually, from yaml 1.1 to yaml 1.2 it looks like they changed from ```0777``` format for octal to ```0o777``` format. 

But in any event, that is not relevant for converting from a yaml string. When the yaml scalar value is a string, and all digits, we can enforce only octal conversion of the number rather than relying on automatic conversion. (If we decide to allow digits in the string version at all. I guess there is an argument to be made that we shouldn't bother.)

**Comment by nick-enoent** [comment]
> > > I python, we don't even necessarily know that the number was specified in octal or decimal, we just have an integer.
> > 
> > 
> > No true. Python octal is specified with a leading '0o'. That said this is not the most user-friendly.
> 
> I think you may be confusing data types with automatic string conversions. Octal is not a data type in python. int() is a data type. The generic yaml parsing library will see "yaml number" and convert it into an "int()" in python. LDMS's python code would not see "octal" when the the yaml scalar value is a number.
> 
> > That's fine, but we need to decide what the configuration file export looks like.
> 
> I'm not suggesting any change to that. I think Nick's perm_handler() code always generates the LDMS's octal string representation, and I think it is a good way to handle that.

Yep, this is why I implemented permissions in this manner. The generic Python YAML parser will automatically convert an octal number in YAML to its integer equivalent, so while Python may support octals with "0o", the octal defined in the YAML never ends up being read as such by the Python code.

The current implementation supports octal numbers in string format, but that may be poorly documented. I'm fine with retooling the unix-like code to represent Chris' suggestion. I'll go ahead and implement that for now, and leave the string octals as is. 

I think it may be worthwhile to add a type check of the permissions once it's parsed on the python side to ensure it's a string. If not, return a type error. This _should_ help limit confusion.

**Comment by tom95858** [comment]
I'm proposing that the perm is always a string, bypassing the YAML conversion.

**Comment by baallan** [comment]
> I'm proposing that the perm is always a string, bypassing the YAML conversion.

Yes, please, please, please!
and maybe seriously consider doing the same for any option which is a time expected to support units; non-unitted times should always be microseconds.

**Comment by tom95858** [comment]
> Oh, actually, from yaml 1.1 to yaml 1.2 it looks like they changed from `0777` format for octal to `0o777` format.
> 
> But in any event, that is not relevant for converting from a yaml string. When the yaml scalar value is a string, and all digits, we can enforce only octal conversion of the number rather than relying on automatic conversion. (If we decide to allow digits in the string version at all. I guess there is an argument to be made that we shouldn't bother.)

We're going to implement and test what has already been described. The perm string will **only** be a string. Anything else will result in an error. We're not going to endlessly negotiate this thing. This will give us complete control over the parsing and error reporting regardless of which YAML parser version is installed.

**Comment by nick-enoent** [comment]
Resolved with 18112af 5120368 and 9c18084
